File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    int min=Integer.MIN_VALUE,max=Integer.MAX_VALUE;
    public int myAtoi(String input) {
        // Your code goes here
        String s=input.trim();
        if(s.isEmpty()) return 0;
        int i=0;
        int sign=1;
        if(s.charAt(i)=='+' || s.charAt(i)=='-'){
            sign=(s.charAt(i)=='+')?1:-1;
            i++;
        }
        return helper(s,i,0,sign);
    }
    public int helper(String s,int i,long num,int sign){
        if(i>=s.length() || !Character.isDigit(s.charAt(i))){
            return (int)(sign*num);
        }
        char ch=s.charAt(i);
        num=num*10+ch-'0';
       if(sign*num<=min) return min;
       if(sign*num>=max) return max;
        return helper(s,i+1,num,sign);

    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s=s.strip()
        i=0
        sign=1
        if len(s)==0: return 0
        if s[i]== '+':
            sign=1
            i+=1
        elif s[i] == '-':
            sign=-1
            i+=1
        INT_MIN = -2**31
        INT_MAX = 2**31 - 1

        def helper(str,i,num,sign):
            if i >= len(s) or not s[i].isdigit():
                return sign * num

    # Update num
            num = num * 10 + int(s[i])

    # Clamp if overflow
            if sign * num <= INT_MIN: return INT_MIN
            if sign * num >= INT_MAX: return INT_MAX

            # Recurse
            return helper(s, i + 1, num, sign)

        return helper(str,i,0,sign)
